processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java

```diff
@@ -0,0 +1,250 @@
+  {
+    return DoubleFirstAggregatorFactory.TIME_COMPARATOR.compare(lhs, rhs) <= 0 ? lhs : rhs;
+  }
+
+  @Override
+  public AggregatorFactory getCombiningFactory()
+  {
+    return new LongFirstAggregatorFactory(name, name)
+    {
+      @Override
+      public Aggregator factorize(ColumnSelectorFactory metricFactory)
+      {
+        final ObjectColumnSelector selector = metricFactory.makeObjectColumnSelector(name);
+        return new LongFirstAggregator(name, null, null)
+        {
+          @Override
+          public void aggregate()
+          {
+            SerializablePair<Long, Long> pair = (SerializablePair<Long, Long>) selector.get();
+            if (pair.lhs < firstTime) {
+              firstTime = pair.lhs;
+              firstValue = pair.rhs;
+            }
+          }
+        };
+      }
+
+      @Override
+      public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
+      {
+        final ObjectColumnSelector selector = metricFactory.makeObjectColumnSelector(name);
+        return new LongFirstBufferAggregator(null, null)
+        {
+          @Override
+          public void aggregate(ByteBuffer buf, int position)
+          {
+            SerializablePair<Long, Long> pair = (SerializablePair<Long, Long>) selector.get();
+            long firstTime = buf.getLong(position);
+            if (pair.lhs < firstTime) {
+              buf.putLong(position, pair.lhs);
+              buf.putLong(position + Longs.BYTES, pair.rhs);
+            }
+          }
+        };
+      }
+    };
+  }
+
+  @Override
+  public AggregatorFactory getMergingFactory(AggregatorFactory other) throws AggregatorFactoryNotMergeableException
+  {
+    if (other.getName().equals(this.getName()) && this.getClass() == other.getClass()) {
+      return getCombiningFactory();
+    } else {
+      throw new AggregatorFactoryNotMergeableException(this, other);
+    }
+  }
+
+  @Override
+  public List<AggregatorFactory> getRequiredColumns()
+  {
+    return Arrays.<AggregatorFactory>asList(new LongFirstAggregatorFactory(fieldName, fieldName));
+  }
+
+  @Override
+  public Object deserialize(Object object)
+  {
+    Map map = (Map) object;
+    return new SerializablePair<>(((Number) map.get("lhs")).longValue(), ((Number) map.get("rhs")).longValue());
+  }
+
+  @Override
+  public Object finalizeComputation(Object object)
+  {
+    return ((SerializablePair<Long, Long>) object).rhs;
+  }
+
+  @Override
+  @JsonProperty
+  public String getName()
+  {
+    return name;
+  }
+
+  @JsonProperty
+  public String getFieldName()
+  {
+    return fieldName;
+  }
+
+  @Override
+  public List<String> requiredFields()
+  {
+    return Arrays.asList(Column.TIME_COLUMN_NAME, fieldName);
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] fieldNameBytes = StringUtils.toUtf8(fieldName);
+
+    return ByteBuffer.allocate(2 + fieldNameBytes.length).put(CACHE_TYPE_ID).put(fieldNameBytes).put((byte)0xff).array();
```

Comment:
nit: can you put the chained function calls on separate lines? I think it makes it easier to see the composition of the cache key that way